expression/builtin_compare.go

```diff
@@ -1383,45 +1393,141 @@ func (c *compareFunctionClass) refineArgsByUnsignedFlag(ctx sessionctx.Context,
-				// `unsigned_col > -1` equals to `1 > 0`,
-				// `unsigned_col <= -1` equals to `1 <= 0`,
-				// `unsigned_col >= 0` equals to `1 >= 0`,
-				// `unsigned_col == -1` equals to `1 == 0`,
-				// `unsigned_col != -1` equals to `1 != 0`,
-				// `unsigned_col <=> -1` equals to `1 <=> 0`,
-				// so we can replace the column argument with `1`, and the other constant argument with `0`.
-				args[i], args[1-i] = NewOne(), NewZero()
-				return args
+			}
+			if !mysql.HasUnsignedFlag(con.RetType.Flag) && !mysql.HasUnsignedFlag(col.RetType.Flag) {
+				tp := col.RetType.Tp
+				lowerBound, upperBound := types.IntergerSignedLowerBound(tp), types.IntergerSignedUpperBound(tp)
+				lowerOverflow, upperOverflow := v < lowerBound, v > upperBound
+				op := c.op
+				if i == 1 {
+					op = symmetricOp[c.op]
+				}
+				if upperOverflow { // the constant value always greater than the column value
+					finalCol, finalCon, succ = c.determineConstComparisonResult(op, colHasNotNullFlag, true, true)
+				} else if lowerOverflow { // the constant value always less than the column value
+					finalCol, finalCon, succ = c.determineConstComparisonResult(op, colHasNotNullFlag, true, false)
+				} else if v == lowerBound || v == upperBound {
+					finalCol, finalCon, succ = c.determineConstComparisonResult(op, colHasNotNullFlag, false, v == upperBound)
+				}
+			}
+			if succ {
+				args[i], args[1-i] = finalCol, finalCon
 			}
 		}
 	}
 	return args
 }
 
+// determineConstComparisonResult used to convert the result of the comparison into constant true or constant false if necessary.
+// The comparison is like `column <op> constant`. colHasNotNullFlag means whether the column has not null flag.
+// isOverflow is used to indicate whether the value of constant exceeds the range of the column.
+// If `isOverflow = false` means the value of constant is equal to upper/ lower bound of column.
+// upper is used to indicate whether it is the upper boundary.
+func (c *compareFunctionClass) determineConstComparisonResult(op opcode.Op, colHasNotNullFlag bool, isOverflow bool, upper bool) (col Expression, con Expression, succ bool) {
+	if isOverflow {
+		if upper { // upperOverflow
+			// The value of constant bigger than the upper bound of column.
+			// For example: column is tinyint type, the range of tinyint is [-128, 127]; the value constant are 200;
+			// 200 > 128 => The value of constant(200) is bigger than the upper bound of column(127). => upperOverflow
+			if op == opcode.EQ || op == opcode.NullEQ || op == opcode.GE || op == opcode.GT {
+				// 127(max) = 200 => always false => 0(column) = 1(constant)
+				// 127(max) <=> 200 => always false => 0(column) <=> 1(constant)
+				// 127(max) >= 200 => always false => 0(column) >= 1(constant)
+				// 127(max) > 200 => always false => 0(column) > 1(constant)
+				col, con, succ = NewZero(), NewOne(), true
+			} else if (op == opcode.NE || op == opcode.LE || op == opcode.LT) && colHasNotNullFlag {
```

Comment:
Seems that this optimization would be broken when column contains `NULL` values, e.g, for `col <= 200`, even if `col` is defined as `tinyint`, this filter may return `NULL` if the value of `col` is NULL, instead of the `true` here.

Reply:
> Seems that this optimization would be broken when column contains `NULL` values, e.g, for `col <= 200`, even if `col` is defined as `tinyint`, this filter should return `NULL` if the value of `col` is NULL, instead of the `true` here.

I checked whether the column contains `Not Null floag`(Use the `colHasNotNullFlag` variable here ), will it help?